pkg/certificate/controller.go

```diff
@@ -0,0 +1,95 @@
+package certificate
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Add creates a new Node Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (m *Manager) Add(mgr manager.Manager) error {
+	return m.add(mgr, m)
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func (m *Manager) add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("certificate-controller", mgr, controller.Options{Reconciler: m})
+	if err != nil {
+		return errors.Wrap(err, "failed instanciating certificate controller")
+	}
+
+	// Watch only events for selected m.webhookName
+	onEventForThisWebhook := predicate.Funcs{
```

Comment:
so, every change in related secret/{mutating,validating}WebhookConfiguration will bring up the reconcile method.
Is cert expiration included as an event of such a kind that will cause the Reconcile to operate? what event specifically?

Reply:
So, expiration is handled by changing `RequeueAfter` at `Reconcile` return if something not related to expiration happends and Reconcile is call, reconcile verify that we are not after deadline and just reset `RequeueAfter` without `rotate`.

Reply:
ack, cool!